src/hotspot/share/gc/g1/g1ConcurrentMark.hpp

```diff
@@ -136,10 +136,92 @@ class G1CMMarkStack {
     G1TaskQueueEntry data[EntriesPerChunk];
   };
 
-  size_t _max_chunk_capacity;    // Maximum number of TaskQueueEntryChunk elements on the stack.
+  class ChunkAllocator {
+    // The chunk allocator relies on a growable array data structure that allows resizing without the
+    // need to copy existing items. The fundamental approach involves organizing the array into chunks,
+    // essentially creating an "array of arrays"; referred to as buckets in this implementation. To
+    // facilitate efficient indexing, the size of the first chunk is set to a power of 2. This choice
+    // allows for quick conversion of an array index into a bucket index and the corresponding offset
+    // within the bucket. Additionally, each new chunk added to the growable array doubles the capacity of
+    // the growable array.
+    //
+    // Illustration of the Growable Array data structure.
+    //
+    //        +----+        +----+----+
+    //        |    |------->|    |    |
+    //        +----+        +----+----+
```

Comment:
```suggestion
    //        +----+
    //        |    |        +----+----+
    //        |    |------->|    |    |
    //        +----+        +----+----+
```
(Not sure why the first bucket has smaller height than the others)